Emit DIE's size in bits when size is not a multiple of 8

augusto2112 opened the pull request:
The existing code will always round down the size in bits when
calculating the size in bytes (for example, a types with 1-7 bits will
be emitted as 0 bytes long). Fix this by emitting the size in bits if
the size is not aligned to a byte.

dwblaikie commented:
Please include a test case (& perhaps soem motivation - are there any targets that actually have non-byte-size types? Perhaps we should disallow that in the verifier instead?)

augusto2112 commented:
Just added some tests!

Motivation wise: Swift has builtin types which are non-byte-size. For example, the builtin that the standard boolean wraps is 1 bit long.

dwblaikie commented:
> Just added some tests!

Thanks!

> Motivation wise: Swift has builtin types which are non-byte-size. For example, the builtin that the standard boolean wraps is 1 bit long.

Hmm, I'm still not sure about this - if the DWARF is describing it as being 1 byte long, presumably that ABI has to actually allocate a byte for this type? So in what sense is the type only 1 bit long?

adrian-prantl commented:
> > Just added some tests!
> 
> Thanks!
> 
> > Motivation wise: Swift has builtin types which are non-byte-size. For example, the builtin that the standard boolean wraps is 1 bit long.
> 
> Hmm, I'm still not sure about this - if the DWARF is describing it as being 1 byte long, presumably that ABI has to actually allocate a byte for this type? So in what sense is the type only 1 bit long?

Swift allows packing values in aggregates under certain circumstances. For example the type `Bool?` a.k.a. `Optional<Bool>` stores the discriminator bit of the Optional enum type in bit #1 and the Bool in bit #0 of the same type. For these applications the bit size is relevant.

dwblaikie commented:
> > > Just added some tests!
> > 
> > 
> > Thanks!
> > > Motivation wise: Swift has builtin types which are non-byte-size. For example, the builtin that the standard boolean wraps is 1 bit long.
> > 
> > 
> > Hmm, I'm still not sure about this - if the DWARF is describing it as being 1 byte long, presumably that ABI has to actually allocate a byte for this type? So in what sense is the type only 1 bit long?
> 
> Swift allows packing values in aggregates under certain circumstances. For example the type `Bool?` a.k.a. `Optional<Bool>` stores the discriminator bit of the Optional enum type in bit #1 and the Bool in bit #0 of the same byte. For these applications the bit size is relevant.

Ah - this sounds like a different thing though - like structs with tail padding that derived classes can use, we don't describe the base class with tail padding as being smaller.

I guess maybe a different direction to discuss: I think it's useful for the DWARF output to match the IR metadata. The fact that we encode the size in bits instead of bytes is perhaps a mistake/certainly opens things up to potential variation between the bit size encoding and the byte size encoding.. 

But what does the non-rounded-up/storage bit size being in the IR metadata provide, if the result in the DWARF is rounded up anyway? Are we going to encode both the storage size and the bit size separately? Using what attributes? and/or Is the IR metadata consumed by something else that uses the bit-size?

adrian-prantl commented:
@augusto2112 Would emitting a `DW_AT_bit_size` be a reasonable thing to do here?

augusto2112 commented:
@adrian-prantl would we always emit bit sizes in that case?

adrian-prantl commented:
I'm thinking: `if (sizeInBits % 8) emitBitSize(sizeInBits) else emitSize(sizeInBits/8)`

dwblaikie commented:
I'm still a bit confused - it seems like the actual size (like what happens if you create an array of these things? I assume it's size is actually in bytes, right?) is still in bytes. But like tail padding in C++ - sometimes derived objects can use some of that padding/size for other stuff too. It doesn't mean the base class is smaller - it's got to be a whole byte for alignment reasons, etc, but the padding bits can be reused.

I think it'd be suitable to model tihs situation more like that situation - where the size is the storage size, but there's some way to describe the padding bits and allow reuse of those.

augusto2112 commented:
> I'm still a bit confused - it seems like the actual size (like what happens if you create an array of these things? I assume it's size is actually in bytes, right?) is still in bytes. But like tail padding in C++ - sometimes derived objects can use some of that padding/size for other stuff too. It doesn't mean the base class is smaller - it's got to be a whole byte for alignment reasons, etc, but the padding bits can be reused.

Yes, the actual size is still in bytes. I think I will emit the correct size in the front end instead, but I think the current implementation that silently rounds down the bits to the bytes is wrong, so we should do something about that separately.

dwblaikie commented:
> > I'm still a bit confused - it seems like the actual size (like what happens if you create an array of these things? I assume it's size is actually in bytes, right?) is still in bytes. But like tail padding in C++ - sometimes derived objects can use some of that padding/size for other stuff too. It doesn't mean the base class is smaller - it's got to be a whole byte for alignment reasons, etc, but the padding bits can be reused.
> 
> Yes, the actual size is still in bytes. I think I will emit the correct size in the front end instead, but I think the current implementation that silently rounds down the bits to the bytes is wrong, so we should do something about that separately.

Yeah - not sure why we even encode it as bits. Maybe we use the bit size for bitfields? A verifier check that the size is a multiple of 8 in cases other than bitfields might be suitable?

augusto2112 commented:
@dwblaikie I talked with Adrian, his idea is to emit the size in bits whenever it does not cleanly fit in a byte. His point (and I agree) is that there may be tools that can use this bit size, and if we always round up when emitting the DWARF we will lose this information, with no way to recover it. I updated the patch to implement this behavior.

Michael137 commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:
```diff
@@ -2210,9 +2216,16 @@ bool DWARFASTParserClang::CompleteRecordType(const DWARFDIE &die,
+          die.GetAttributeValueAsUnsigned(DW_AT_byte_size, UINT64_MAX);
+
+      if (byte_size != UINT64_MAX)
+        layout_info.bit_size = byte_size;
```
How come we don't need to do the `* 8` here to go from bytes to bits?

augusto2112 replied:
Thanks for catching this.

Michael137 commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:
```diff
@@ -2866,8 +2879,12 @@ void DWARFASTParserClang::ParseSingleMember(
+    parent_bit_size = parent_byte_size * 8;
+  else
+    parent_bit_size =
+        parent_die.GetAttributeValueAsUnsigned(DW_AT_bit_size, UINT64_MAX);
```
Thoughts on extracting this block into a helper? Then you can use it above as well.

Michael137 commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:
```diff
@@ -2866,8 +2879,12 @@ void DWARFASTParserClang::ParseSingleMember(
+    parent_bit_size = parent_byte_size * 8;
+  else
+    parent_bit_size =
+        parent_die.GetAttributeValueAsUnsigned(DW_AT_bit_size, UINT64_MAX);
```
Do we already have tests in LLDB that cover this?

augusto2112 replied:
Hmm, I'll definitely have a downstream test that exercises this behavior, but I don't think on the clang side this would be easy to test, do you have any suggestions on how to do that?

Michael137 replied:
DWARFASTParserClang has unittests where we sometimes use Yaml2Obj and directly invoke the parser. But it's quite finicky. Let me know if you run into troubles with it

augusto2112 committed:
Emit DIE's size in bits when size is not a multiple of 8

The existing code will always round down the size in bits when
calculating the size in bytes (for example, a types with 1-7 bits will
be emitted as 0 bytes long). Fix this by emitting the size in bits if
the size is not aligned to a byte.

dwblaikie commented:
> @dwblaikie I talked with Adrian, his idea is to emit the size in bits whenever it does not cleanly fit in a byte. His point (and I agree) is that there may be tools that can use this bit size, and if we always round up when emitting the DWARF we will lose this information, with no way to recover it. I updated the patch to implement this behavior.

Yeah - ultimately it's pretty much up to you folks how you encode Swift and ObjC things - so take anything I'm saying as advice but not requirement.

But it feels a bit at-odds with precedent in the way other things that seem similar to me are encoded in DWARF already, and precedent is pretty much most/all we have to go with in DWARF.

I guess one question that might be relevant - does Swift have something like `sizeof` and what result does it give for these sort of types with bits to spare?

But like I said - it seems like structs with tail padding are similar to this situation - we still describe the whole size of the struct, because that's used for creating arrays of instances, ABI passing, etc. But the tail padding can still be used, in certain circumstances, when laying out a derived class. We encode this as the POD-ness of the type, and so if you wanted to create a class that derived from one described in DWARF you could do so & would know whether or not to put the derived class's members into the tail padding of the base or not.

augusto2112 commented:
> I guess one question that might be relevant - does Swift have something like sizeof and what result does it give for these sort of types with bits to spare?

You can't actually use that with these types as these are special compiler builtin types which aren't actually accessible in source code.

> But like I said - it seems like structs with tail padding are similar to this situation - we still describe the whole size of the struct, because that's used for creating arrays of instances, ABI passing, etc. But the tail padding can still be used, in certain circumstances, when laying out a derived class. We encode this as the POD-ness of the type, and so if you wanted to create a class that derived from one described in DWARF you could do so & would know whether or not to put the derived class's members into the tail padding of the base or not.

I understand the rationale of basing this on precedent, but in this case in this case we should break from it for two reasons:

- DW_AT_BIT_SIZE is already a standardized attribute in Dwarf that fits this use case.
- Round up to the nearest byte would lose information, which can be kept with fairly minimal downsides in my opinion.

dwblaikie commented:
> > I guess one question that might be relevant - does Swift have something like sizeof and what result does it give for these sort of types with bits to spare?
> 
> You can't actually use that with these types as these are special compiler builtin types which aren't actually accessible in source code.

Perhaps observable indirectly?
 
> > But like I said - it seems like structs with tail padding are similar to this situation - we still describe the whole size of the struct, because that's used for creating arrays of instances, ABI passing, etc. But the tail padding can still be used, in certain circumstances, when laying out a derived class. We encode this as the POD-ness of the type, and so if you wanted to create a class that derived from one described in DWARF you could do so & would know whether or not to put the derived class's members into the tail padding of the base or not.
> 
> I understand the rationale of basing this on precedent, but in this case in this case we should break from it for two reasons:
> 
> * DW_AT_BIT_SIZE is already a standardized attribute in Dwarf that fits this use case.

I'm arguing it doesn't fit it particularly well. We use it for bit fields - which are pretty special, for instance, but it seems like this thing isn't quite like that - it does have a whole byte size (if you allocated an array of them, for instance, I'm guessing they're a byte each, right?) but then has some padding bits that can be reused in some circumstances? That's why I'm saying it seems like it fits more closely to the struct padding representation.

> * Round up to the nearest byte would lose information, which can be kept with fairly minimal downsides in my opinion.

Seems like it'd still need to be special cased, right? The consumer would go "oh, this has a bit size, but if we want an array of them, or to allocate them for ABI purposes, etc, I have to round it up to the nearest byte"? or something like that.

Some pointers to documentation about these types, and the range of uses/instances there are might be handy (like is this a general concept? Or is it only one type that uses this (`bool` equivalent, with 7 padding bytes unused) or a class of types (a small finite list of them? Unbounded (like if I put a bool in my custom struct - does my custom struct end up with a bit size too?))

adrian-prantl commented:
> I'm arguing it doesn't fit it particularly well. We use it for bit fields - which are pretty special, for instance, but it seems like this thing isn't quite like that - it does have a whole byte size (if you allocated an array of them, for instance, I'm guessing they're a byte each, right?) but then has some padding bits that can be reused in some circumstances? That's why I'm saying it seems like it fits more closely to the struct padding representation.

Swift is really clever at packing at packing aggregate types. For example, the discriminator bits for enums are always stored in unused bits of the payload type. For a contrived example, the type Optional<Optional<Bool>> has a size of 3 bits.
